File: CIMWeb/src/main/java/ch/ninecode/cim/cimweb/GridLabExport.java
package ch.ninecode.cim.cimweb;

import java.io.PrintWriter;
import java.io.StringWriter;

import javax.annotation.Resource;
import javax.ejb.Stateless;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.resource.ResourceException;
import javax.resource.cci.Connection;
import javax.resource.cci.Interaction;
import javax.resource.cci.MappedRecord;

import ch.ninecode.cim.connector.CIMConnectionFactory;
import ch.ninecode.cim.connector.CIMConnectionSpec;
import ch.ninecode.cim.connector.CIMInteractionSpec;
import ch.ninecode.cim.connector.CIMInteractionSpecImpl;
import ch.ninecode.cim.connector.CIMMappedRecord;

import ch.ninecode.gl.Line;

@Stateless
@Path("/GridLabExport/{file}")
public class GridLabExport
{
    @Resource (lookup="openejb:Resource/CIMConnector.rar")
    CIMConnectionFactory factory;

    /**
     * Build a connection specification used by all the tests.
     * @return
     */
    CIMConnectionSpec remoteConfig ()
    {
        CIMConnectionSpec ret;

        ret = new CIMConnectionSpec ();
        ret.setUserName ("derrick"); // not currently used
        ret.setPassword ("secret"); // not currently used
        ret.getProperties ().put ("spark.driver.memory", "1g");
        ret.getProperties ().put ("spark.executor.memory", "4g");

        return (ret);
    }

    @SuppressWarnings ("unchecked")
    @GET
    @Path("{p:/?}{item:((.*)?)}")
    public Response GetGridLABExport (@PathParam("file") String filename, @PathParam("item") String item)
    {
        String transformer = (null != item && !item.equals ("")) ? item : null;
        StringBuffer out = new StringBuffer ();
        if (null != factory)
        {
            Connection connection;
            try
            {
                connection = factory.getConnection (remoteConfig ());
                if (null != connection)
                {
                    try
                    {
                        String full_file = "hdfs://sandbox:8020/data/" + filename + ".rdf";
                        final CIMInteractionSpecImpl spec = new CIMInteractionSpecImpl ();
                        spec.setFunctionName (CIMInteractionSpec.GET_STRING_FUNCTION);
                        final MappedRecord input = factory.getRecordFactory ().createMappedRecord (CIMMappedRecord.INPUT);
                        input.setRecordShortDescription ("record containing the file name and class and method to run");
                        input.put ("filename", full_file);

                        // set up the method call details for the CIMConnector
                        Line line = new Line ();
                        input.put ("class", line.getClass ().getName ());
                        String jar = factory.JarPath (line);
                        if (null != jar)
                            input.put ("jars", jar);
                        if (null == transformer)
                            input.put ("method", "prepare");
                        else
                            input.put ("method", "export");
                        if (null != transformer)
                            input.put ("transformer", transformer);
                        final MappedRecord output = factory.getRecordFactory ().createMappedRecord (CIMMappedRecord.OUTPUT);
                        output.setRecordShortDescription ("the results of the gridlab conversion");
                        final Interaction interaction = connection.createInteraction ();
                        if (interaction.execute (spec, input, output))
                        {
                            String result = output.get ("result").toString ();
                            out.append (result);
                        }
                        interaction.close ();
                        connection.close ();
                    }
                    catch (ResourceException resourceexception)
                    {
                        out.append ("ResourceException on interaction");
                        out.append ("\n");
                        StringWriter string = new StringWriter ();
                        PrintWriter writer = new PrintWriter (string);
                        resourceexception.printStackTrace (writer);
                        out.append (string.toString ());
                        writer.close ();
                    }
                    finally
                    {
                        try
                        {
                            connection.close ();
                        }
                        catch (ResourceException resourceexception)
                        {
                            out.append ("ResourceException on close");
                            out.append ("\n");
                            StringWriter string = new StringWriter ();
                            PrintWriter writer = new PrintWriter (string);
                            resourceexception.printStackTrace (writer);
                            out.append (string.toString ());
                            writer.close ();
                        }
                    }
                }
            }
            catch (ResourceException exception)
            {
                out.append ("ResourceException");
                out.append ("\n");
                StringWriter string = new StringWriter ();
                PrintWriter writer = new PrintWriter (string);
                exception.printStackTrace (writer);
                out.append (string.toString ());
                writer.close ();
            }
        }
        return (Response.ok (out.toString (), MediaType.APPLICATION_OCTET_STREAM)
            .header ("content-disposition", "attachment; filename =" + "gridlabd.glm")
            .build ());
    }
}

File: CIMWeb/src/main/scala/ch/ninecode/cim/cimweb/GridLABExport.scala
package ch.ninecode.cim.cimweb

import java.util.logging.Logger
import javax.ejb.Stateless
import javax.ws.rs.GET
import javax.ws.rs.core.MediaType
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.resource.ResourceException
import javax.ws.rs.core.Response

import ch.ninecode.cim.connector.CIMFunction
import ch.ninecode.cim.connector.CIMInteractionSpec
import ch.ninecode.cim.connector.CIMInteractionSpecImpl
import ch.ninecode.cim.connector.CIMMappedRecord

@Stateless
@Path ("gridlab/")
class GridLABExport extends RESTful
{
    import GridLABExport._

    // get a righteous string for a glm filename
    def glm_name (simulation: String): String =
    {
        val index = simulation.lastIndexOf ("/")
        val suffix = if (simulation.endsWith (".json")) simulation.length - 4 else simulation.length
        (if (-1 == index) simulation.substring (0, suffix) else simulation.substring (index + 1, suffix)) + ".glm"
    }

    @GET
    @Path ("{simulation:[^;]*}")
    @Produces (Array (MediaType.APPLICATION_JSON))
    def export (
        @PathParam ("simulation") simulation: String // the name of the JSON simulation file on HDFS
        ): Response =
    {
        _Logger.info ("gridlab %s".format (simulation))
        val ret = new RESTfulJSONResult
        val connection = getConnection (ret)
        val response: Response = if (null != connection)
            try
            {
                val spec: CIMInteractionSpec = new CIMInteractionSpecImpl
                spec.setFunctionName (CIMInteractionSpec.EXECUTE_CIM_FUNCTION)
                val input = getInputRecord ("input record containing the function to run")
                // set up the function with parameters
                val gridlab = GridLABExportFunction (if (simulation.startsWith ("/")) simulation else "/" + simulation)
                input.asInstanceOf[map].put (CIMFunction.FUNCTION, gridlab)
                val interaction = connection.createInteraction
                val output = interaction.execute (spec, input)
                if (null == output)
                    throw new ResourceException ("null is not a MappedRecord")
                else
                    if (!output.getClass.isAssignableFrom (classOf [CIMMappedRecord]))
                        throw new ResourceException ("object of class %s is not a MappedRecord".format (output.getClass.toGenericString))
                    else
                    {
                        val record = output.asInstanceOf [CIMMappedRecord]
                        Response.ok (record.get (CIMFunction.RESULT).asInstanceOf [String], MediaType.APPLICATION_OCTET_STREAM)
                            .header ("content-disposition", "attachment; filename=%s".format (glm_name (simulation)))
                            .build
                    }
            }
            catch
            {
                case resourceexception: ResourceException =>
                    ret.setResultException (resourceexception, "ResourceException on interaction")
                    Response.serverError ().entity (ret.message).build
            }
            finally
                try
                    connection.close ()
                catch
                {
                    case resourceexception: ResourceException =>
                        ret.setResultException (resourceexception, "ResourceException on close")
                        Response.serverError ().entity (ret.message).build
                }
        else
            Response.status (Response.Status.SERVICE_UNAVAILABLE).entity ("could not get connection").build

        response
    }
}

object GridLABExport
{
    val LOGGER_NAME: String = GridLABExport.getClass.getName
    val _Logger: Logger = Logger.getLogger (LOGGER_NAME)
}
